### PR TITLE
test: followup update chainnotifier mock to fwd notifications for matching txs

### PR DESCRIPTION
**Pull request by bhandras**
This is the remainder of the previous PR (https://github.com/lightninglabs/loop/pull/174) which was left out by my mistake. This is required to make it fully possible to test scenarios where we're watching the chain for different versions of the same transaction (eg. NP2WSH and P2WSH)

### CODE REVIEW

File: test/chainnotifier_mock.go [hunk 1]
```diff
@@ -112,26 +123,25 @@ func (c *mockChainNotifier) RegisterConfirmationsNtfn(ctx context.Context,
...
 		select {
 		case m := <-c.lnd.ConfChannel:
-			if bytes.Equal(m.Tx.TxOut[0].PkScript, pkScript) {
-				select {
-				case confChan <- m:
-				case <-ctx.Done():
+			for _, r := range c.confRegistrations {
+				// Whichever conf notifier catches the confirmation
+				// will forward it to all matching subscibers.
+				if bytes.Equal(m.Tx.TxOut[0].PkScript, r.PkScript) {
```
Comment (joostjager) [inline test/chainnotifier_mock.go:129]: Remove registration. Or not?

File: test/chainnotifier_mock.go [hunk 2]
```diff
@@ -103,7 +105,16 @@ func (c *mockChainNotifier) RegisterConfirmationsNtfn(ctx context.Context,
...
-	confChan := make(chan *chainntnfs.TxConfirmation, 1)
+	reg := &ConfRegistration{
+		PkScript:   pkScript,
+		TxID:       txid,
+		HeightHint: heightHint,
+		NumConfs:   numConfs,
+		ConfChan:   make(chan *chainntnfs.TxConfirmation, 1),
+	}
+
+	c.confRegistrations = append(c.confRegistrations, reg)
```
Comment (joostjager) [inline test/chainnotifier_mock.go:116]: Thread safety?

Comment (bhandras) [reply test/chainnotifier_mock.go:118]: Good catch, I guess I didn't notice this as there weren't concurrent registrations in our current tests, not even after adding native segwit.

File: test/chainnotifier_mock.go (hunk 1 above)
Comment (bhandras) [reply test/chainnotifier_mock.go:135]: Also good suggestion. Pretty simple to remove so why not.

File: test/chainnotifier_mock.go [hunk 3]
```diff
@@ -112,26 +126,35 @@ func (c *mockChainNotifier) RegisterConfirmationsNtfn(ctx context.Context,
...
+			c.Lock()
+			for i := 0; i < len(c.confRegistrations); i++ {
+				r := c.confRegistrations[i]
+
+				// Whichever conf notifier catches the confirmation
+				// will forward it to all matching subscibers.
+				if bytes.Equal(m.Tx.TxOut[0].PkScript, r.PkScript) {
+					// Unregister the "notifier".
+					c.confRegistrations = append(
+						c.confRegistrations[:i], c.confRegistrations[i+1:]...,
```
Comment (joostjager) [inline test/chainnotifier_mock.go:138]: Isn't `i+1` a crash if it is the last reg? Probably not as I see similar examples on internet

File: test/chainnotifier_mock.go [hunk 4]
```diff
@@ -112,26 +126,35 @@ func (c *mockChainNotifier) RegisterConfirmationsNtfn(ctx context.Context,
...
+					)
+					i--
```
Comment (joostjager) [inline test/chainnotifier_mock.go:140]: I am thinking about a better way to go about this, but can't think of it. It works.

File: test/chainnotifier_mock.go (hunk 3 above)
Comment (bhandras) [reply test/chainnotifier_mock.go:138]: it's valid (https://play.golang.org/p/TXP0B4ZtCNw)

File: test/chainnotifier_mock.go (hunk 4 above)
Comment (bhandras) [reply test/chainnotifier_mock.go:140]: linked list could fit better

**Pull request merged**